File: Data Structures & Algorithms/subsets-ii/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsetsWithDup(nums) {
        function backtracking(ind) {
            if (ind === nums.length) {
                res.push([...subset]);
                return;
            }

            subset.push(nums[ind]);
            backtracking(ind+1);
            subset.pop();

            while(ind+1 < nums.length && nums[ind] === nums[ind+1]) {
                ind++;
            };
            backtracking(ind+1);
        }

        const res = [];
        const subset = [];
        nums.sort((a,b) => a-b);
        backtracking(0);
        return res;
    }
}

File: Data Structures & Algorithms/reconstruct-flight-path/submission-1.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        # Need review :(
        # create empty adj list -> (src : [])
        adj = { src:[] for src, dst in tickets }
        
        # create adj list -> (src : [dst list])
        tickets.sort()
        for src, dst in tickets:
            adj[src].append(dst)
        
        res = ["JFK"]
        def dfs(src):
            # base case
            if len(res) == len(tickets) + 1:
                return True
            # 如果現在的這個 node 沒有可以去的下一個地方就 return False
            if src not in adj:
                return False
            
            temp = list(adj[src])
            for i, v in enumerate(temp):
                # backtracking here
                adj[src].pop(i)
                res.append(v)
                if dfs(v): return True
                adj[src].insert(i, v)
                res.pop()
            
            # return False if no valid path
            return False
        
        dfs("JFK")
        return res
